homeassistant/components/mqtt/water_heater.py

```diff
@@ -0,0 +1,438 @@
+"""Support for MQTT water heater devices."""
+from __future__ import annotations
+
+from collections.abc import Callable
+import functools
+import logging
+from typing import Any
+
+import voluptuous as vol
+
+from homeassistant.components import water_heater
+from homeassistant.components.water_heater import (
+    ATTR_OPERATION_MODE,
+    DEFAULT_MAX_TEMP,
+    DEFAULT_MIN_TEMP,
+    STATE_ECO,
+    STATE_ELECTRIC,
+    STATE_GAS,
+    STATE_HEAT_PUMP,
+    STATE_HIGH_DEMAND,
+    STATE_PERFORMANCE,
+    WaterHeaterEntity,
+    WaterHeaterEntityFeature,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import (
+    ATTR_TEMPERATURE,
+    CONF_NAME,
+    CONF_OPTIMISTIC,
+    CONF_PAYLOAD_OFF,
+    CONF_PAYLOAD_ON,
+    CONF_TEMPERATURE_UNIT,
+    CONF_VALUE_TEMPLATE,
+    PRECISION_HALVES,
+    PRECISION_TENTHS,
+    PRECISION_WHOLE,
+    STATE_OFF,
+)
+from homeassistant.core import HomeAssistant, callback
+import homeassistant.helpers.config_validation as cv
+from homeassistant.helpers.entity_platform import AddEntitiesCallback
+from homeassistant.helpers.template import Template
+from homeassistant.helpers.typing import ConfigType, DiscoveryInfoType
+
+from . import subscription
+from .config import DEFAULT_RETAIN, MQTT_BASE_SCHEMA
+from .const import (
+    CONF_ENCODING,
+    CONF_QOS,
+    CONF_RETAIN,
+    DEFAULT_OPTIMISTIC,
+    PAYLOAD_NONE,
+)
+from .debug_info import log_messages
+from .mixins import MQTT_ENTITY_COMMON_SCHEMA, MqttEntity, async_setup_entry_helper
+from .models import (
+    MqttCommandTemplate,
+    MqttValueTemplate,
+    PublishPayloadType,
+    ReceiveMessage,
+    ReceivePayloadType,
+)
+from .util import get_mqtt_data, valid_publish_topic, valid_subscribe_topic
+
+_LOGGER = logging.getLogger(__name__)
+
+DEFAULT_NAME = "MQTT Water Heater"
+
+CONF_CURRENT_TEMP_TEMPLATE = "current_temperature_template"
+CONF_CURRENT_TEMP_TOPIC = "current_temperature_topic"
+CONF_MODE_COMMAND_TEMPLATE = "mode_command_template"
+CONF_MODE_COMMAND_TOPIC = "mode_command_topic"
+CONF_MODE_LIST = "modes"
+CONF_MODE_STATE_TEMPLATE = "mode_state_template"
+CONF_MODE_STATE_TOPIC = "mode_state_topic"
+CONF_PRECISION = "precision"
+CONF_TEMP_COMMAND_TEMPLATE = "temperature_command_template"
+CONF_TEMP_COMMAND_TOPIC = "temperature_command_topic"
+CONF_TEMP_STATE_TEMPLATE = "temperature_state_template"
+CONF_TEMP_STATE_TOPIC = "temperature_state_topic"
+CONF_TEMP_INITIAL = "initial"
+CONF_TEMP_MAX = "max_temp"
+CONF_TEMP_MIN = "min_temp"
```

Comment:
I think we should move these contants to `const.py` if they are used both in `climate` and `water_heater`

Reply:
Would you like me to replace the ones in `climate.py` as well? Or save for another PR?

Reply:
Yes please, since they are being reused.